src/compiler/checker.ts

```diff
@@ -4722,6 +4846,1104 @@ namespace ts {
+                        [...indexSignatures, ...staticMembers, ...constructors, ...members]
+                    ), symbol.declarations && filter(symbol.declarations, d => isClassDeclaration(d) || isClassExpression(d))[0]), modifierFlags);
+                }
+
+                // Synthesize declarations for a symbol - might be an Interface, a Class, a Namespace, a Type, a Variable (const, let, or var), an Alias
+                // or a merge of some number of those.
+                // An interesting challenge is ensuring that when classes merge with namespaces and interfaces, is keeping
+                // each symbol in only one of the representations
+                // Also, synthesizing a default export of some kind
+                // If it's an alias: emit `export default ref`
+                // If it's a property: emit `export default _default` with a `_default` prop
+                // If it's a class/interface/function: emit a class/interface/function with a `default` modifier
+                // These forms can merge, eg (`export default 12; export default interface A {}`)
+                function serializeSymbolWorker(symbol: Symbol, isPrivate: boolean, propertyAsAlias: boolean) {
+                    const symbolName = unescapeLeadingUnderscores(symbol.escapedName);
+                    const isDefault = symbol.escapedName === InternalSymbolName.Default;
+                    if (isStringANonContextualKeyword(symbolName) && !isDefault) {
+                        // Oh no. We cannot use this symbol's name as it's name... It's likely some jsdoc had an invalid name like `export` or `default` :(
+                        context.encounteredError = true;
+                        // TODO: Issue error via symbol tracker?
+                        return; // If we need to emit a private with a keyword name, we're done for, since something else will try to refer to it by that name
+                    }
+                    const needsPostExportDefault = isDefault && (
+                           !!(symbol.flags & SymbolFlags.ExportDoesNotSupportDefaultModifier)
+                        || (!!(symbol.flags & SymbolFlags.Function) && !!length(getPropertiesOfType(getTypeOfSymbol(symbol))))
+                    ) && !(symbol.flags & SymbolFlags.Alias); // An alias symbol should preclude needing to make an alias ourselves
+                    if (needsPostExportDefault) {
+                        isPrivate = true;
+                    }
+                    const modifierFlags = (!isPrivate ? ModifierFlags.Export : 0) | (isDefault && !needsPostExportDefault ? ModifierFlags.Default : 0);
+                    if (symbol.flags & SymbolFlags.Function) {
+                        const type = getTypeOfSymbol(symbol);
+                        serializeAsFunctionNamespaceMerge(type, symbol, getInternalSymbolName(symbol, symbolName), modifierFlags);
+                    }
+                    if (symbol.flags & SymbolFlags.TypeAlias) {

```

Comment:
I would move any case with more than a couple of lines of code into a separate function so the dispatch stays obvious.